Rust UI: confirm_properties layout

mmilata opened the pull request:
Part of #1922. The biggest change in this PR  is that `Paragraphs` no longer holds its contents in Vec with hardcoded capacity but any data structure implementing a simple trait can be used. I've attempted to put most of the logic into functions that won't be monomorphized in order to save code size. My current solution has the disadvantage that it needs to allocate large buffer on stack (can be moved to globals) when computing paragraph breaks, LMK if you can think of something better. [[UI diff](https://satoshilabs.gitlab.io/-/trezor/trezor-firmware/-/jobs/3053013611/artifacts/master_diff/index.html)]

TODO:
- [x] Put key and value on the same page if possible: [python impl](https://github.com/trezor/trezor-firmware/blob/core/v2.5.2/core/src/trezor/ui/layouts/tt/__init__.py#L814); done: 48becacd8c525fcc6a106c4db3efdddf87a59e9e.
- [x] Test on hardware.
- [ ] Consider renaming `Paragraph` to `Par`/`Para` to avoid mixups with `Paragraphs`.

![cp1](https://user-images.githubusercontent.com/85857/191212363-d242ae11-ee3e-4f33-8474-406c5b035a8d.png) ![cp2](https://user-images.githubusercontent.com/85857/191212356-013d2083-5a35-4507-ab2a-f77ee581f42e.png)

jpochyla commented:
Niiiice

mmilata commented on core/embed/rust/src/ui/model_tt/layout.rs:
```diff
@@ -259,6 +262,23 @@ where
 }
 
+/*
+impl ParagraphSource for Obj
+where
+{
+    fn collect<'a>(&self, sink: &mut impl Extend<Paragraph<&'a str>>) {
+        let mut iter_buf = IterBuf::new();
+        let iter = Iter::try_from_obj_with_buf(*self, &mut iter_buf);
+        let iter = unwrap!(iter);
+        for obj in iter {
+            let text: StrBuffer = unwrap!(obj.try_into());
+            let text = text.as_ref();
+            sink.extend(Some(Paragraph::new(&theme::TEXT_NORMAL, text)))
+        }
+    }
+}
```
I didn't manage to implement taking the paragraphs directly out of `Obj` because I couldn't figure out the lifetimes:
```
error[E0597]: `text` does not live long enough
   --> src/ui/model_tt/layout.rs:275:24
    |
269 |     fn collect<'a>(&self, sink: &mut impl Extend<Paragraph<&'a str>>) {
    |                -- lifetime `'a` defined here
...
275 |             let text = text.as_ref();
    |                        ^^^^^^^^^^^^^ borrowed value does not live long enough
276 |             sink.extend(Some(Paragraph::new(&theme::TEXT_NORMAL, text)))
    |             ------------------------------------------------------------ argument requires that `text` is borrowed for `'a`
277 |         }
278 |     }
    |     - `text` dropped here while still borrowed

```

If we decide not to implement the trait for `Obj` then `ParagraphSource` can be redefined to provide random-access (instead of sequential) which would probably allow us to get rid of the large temporary array.

matejcik replied:
doing this for `Paragraph<StrBuffer>` should be fine, no? the whole point of `StrBuffer` is to steal reasonable lifetimes from under the GC

> redefined to provide random-access (instead of sequential)

we can get random access from uPy lists too. it risks exceptions, but so does using the iterator.

matejcik replied:
what is the status of this? or should we merge as-is and create an issue for `ParagraphSource for Obj`?

mmilata replied:
In #2531 I implemented `ParagraphSource for Vec<Paragraph<Obj>>` - it seemed more straightforward to me to determine the number of paragraphs and their styling at the beginning and only call to uPy to get the contents. If I'm measuring this correctly `ParagraphVecLong<Obj>` has 388 bytes in firmware.

mmilata replied:
And `ParagraphVecLong<StrBuffer>` has 516, might save some memory by converting to `ParagraphVecLong<Obj>`. Quick & dirty way to measure this:

```rust
macro_rules! dump_size {
    ($x:expr $(,)?) => {
        #[allow(unknown_lints, eq_op)]
        const _: [(); 0 - $x as usize] = [];
    };
}
use core::mem;
dump_size!(mem::size_of::<ParagraphVecLong<Obj>>());
```

TychoVrahe commented on core/embed/rust/src/ui/component/text/paragraphs.rs:
```diff
@@ -417,14 +506,31 @@ where
+            return self;
+        }
+        if self.push(paragraph).is_err() {
+            #[cfg(feature = "ui_debug")]
```
could benefit from debug-assertions also here using debug_panic!

TychoVrahe reviewed (commented):
The main issue i have here is with Swipe(Hold)Page and actually not caused in this PR, but nevertheless, it manifests here when testing confirm_properties on device so it seems like a good place to fix it
- the loader in SwipeHoldPage is not positioned correctly and collides with buttons. The place function should take into account height of the button area and probably also BUTTON_SPACING
- the buttons flicker when loader animation is in progres. This is fixed with this commit https://github.com/trezor/trezor-firmware/commit/a4d0ec3926607b54a15a712f2202f1a654a27607, but then the buttons are not redrawn when Hold to confirm is Cancelled. In on_page_change we should probably call request_complete_repaint  of the buttons too.
- when hold to confirm completes, the content is drawn on the remainders of the loader and briefly shown before next actions continue. It should probably just continue with fade out and move to next stuff right away.

mmilata commented:
Rebased to master & fixed the SwipeHoldPage issues in dc675ab825a29c417cfab891b95832a1cd26517a.

Switched `ParagraphSource` to random access in 9cd430fb2c2e04b28fccff05c5b5c19dced2e614.

Implemented key-value-on-same-page in 48becacd8c525fcc6a106c4db3efdddf87a59e9e.

mmilata commented on core/embed/rust/src/ui/model_tt/layout.rs:
```diff
@@ -338,17 +359,66 @@ extern "C" fn new_confirm_blob(n_args: usize, args: *const Obj, kwargs: *mut Map
+            let [key, value, is_mono]: [Obj; 3] = iter_into_objs(para)?;
+
+            if let Some(key) = key.try_into_option::<StrBuffer>()? {
+                paragraphs.add(Paragraph::new(&theme::TEXT_BOLD, key));
```
That was my plan but had no good name for the flag. Implemented in 48becacd8c525fcc6a106c4db3efdddf87a59e9e (only works for two subsequent paragraphs though so not sure about the reusability).

mmilata commented on core/embed/rust/src/ui/component/text/paragraphs.rs:
```diff
@@ -279,12 +381,34 @@ struct PageBreakIterator<'a, T> {
+                return None;
+            } else if let Some(remaining_area) = remaining_area {
+                #[cfg(feature = "ui_debug")]
+                assert!(next_offset.par == offset.par + 1);
```
Fixed in dc675ab825a29c417cfab891b95832a1cd26517a. IIRC some time ago we decided to keep the debug assertions and PYOPT=0 separate.

mmilata commented on core/embed/rust/src/ui/component/text/paragraphs.rs:
```diff
@@ -205,60 +237,73 @@ pub mod trace {
 pub struct Paragraph<T> {
     content: T,
-    layout: TextLayout,
+    style: &'static TextStyle,
```
We'd have to include the lifetime in all types that contain the Paragraphs in them which hurts readability IMO. Let's do it once/if it's needed.

TychoVrahe commented:
One more issue with `SwipeHoldPage` after the fix: when the Hold to confirm button is pressed, or the loader shrinks, the buttons get cleared and redrawn, causing a flicker. Thinking that `SwipeHoldPage` should have its own `Pad` with size of content area returned by `place_get_content_area` which would be passed into `handle_hold_event` instead of `inner.pad`. Note that `place_get_content_area` doesn't currently return content + scrollbar like described in the comment so this needs to be fixed too in order for this to work properly (it returns `content_single_page` which is narrower by `theme::CONTENT_BORDER`, so maybe return `layout.content_single_page.union(layout.scrollbar)`)

mmilata commented:
Good catch! Do you just spot these or do you have some flicker-debugging mode implemented?

I've applied your suggestion in 66698ca and it seems to be working. If there are more similar bugs we might want to re-implement the component using `Maybe` + `PaintOverlapping` which were added some time after the initial `SwipePage` implementation.

matejcik commented on core/embed/rust/src/ui/component/text/paragraphs.rs:
```diff
@@ -279,12 +451,37 @@ struct PageBreakIterator<'a, T> {
     current: Option<PageOffset>,
 }
 
+impl<'a, T: ParagraphSource> PageBreakIterator<'a, T> {
```
this should just be `impl<T> PageBreakIterator<'_, T>` no? as you're not using any of the type arguments in this block, it's basically just a namespace for `dyn_next`

mmilata replied:
Oh, I didn't realize this was possible, fixed in 8a4c769cc2601aa8a3d01bd997a639cc6c5eaa51.

matejcik commented on core/embed/rust/src/ui/model_tt/layout.rs:
```diff
@@ -338,17 +342,72 @@ extern "C" fn new_confirm_blob(n_args: usize, args: *const Obj, kwargs: *mut Map
+                    paragraphs.add(Paragraph::new(&theme::TEXT_MONO, value));
+                } else {
+                    paragraphs.add(Paragraph::new(&theme::TEXT_NORMAL, value));
+                }
```
sidenote: we should violently throw out this flexibility of sometimes specifying only key and sometimes specifying only value. I believe only Cardano is using that, maybe a little bit of UX work is needed on that side.

matejcik commented on core/embed/rust/src/ui/model_tt/layout.rs:
```diff
@@ -1021,6 +1092,17 @@ pub static mp_module_trezorui2: Module = obj_module! {
+    /// def confirm_properties(
+    ///     *,
+    ///     title: str,
+    ///     items: Iterable[Tuple[str, str, bool]],
```
...speaking of, this new type annotation disallows `None`s anyway?

mmilata replied:
Good catch, fixed in 8a4c769cc2601aa8a3d01bd997a639cc6c5eaa51

matejcik reviewed (commented):
LGTM overall, one very small style nit

mmilata commented on core/embed/rust/src/ui/component/text/paragraphs.rs:
```diff
@@ -279,12 +451,37 @@ struct PageBreakIterator<'a, T> {
-where
-    T: AsRef<str>,
-{
+impl<'a, T: ParagraphSource> Iterator for PageBreakIterator<'a, T> {
```
8a4c769cc2601aa8a3d01bd997a639cc6c5eaa51

mmilata commented:
Rebased on master and fixed ui fixtures, will merge if CI passes.
